##### Improting libraries #####
import pandas as pd
from snakemake.utils import min_version
min_version("7.6.1")
##### Improting libraries #####

##### Important thing to run snakemake #####
# By 11th November 2022, snakemake didnt working. I found a solution in a slackoverflow post, suggesting
# to downgrade tabulate package from 0.9 to 0.8.
# https://stackoverflow.com/questions/74226320/snakemake-tutorial-error-attributeerror-str-object-has-no-attribute-name
# So then I executed the command:
# conda activate <env_name>
# conda install tabulate=0.8
# This command installed tabulate version 0.8.10
#
# Install as well snakemake-wrapper-utils
# conda install -c bioconda snakemake-wrapper-utils
##### Important thing to run snakemake #####

##### Workflow guide #####
# RNA-Seq libraries
#   - Assess quality control
#   - Trim adapter sequences and filtering low quality reads - trimmomatic
#       - Map to reference genome
#       - Remove duplicates reads
#       - 
#       - Build up multiqc file
##### Workflow guide #####

##### Loading config file #####
configfile: "../config/config.yaml"
##### Loading config file #####

##### Loading phenodata file #####
phenodata = (
    pd.read_table(config["samples"], dtype={"Filename_R1": str}, delimiter='\t')
    .set_index("Filename_R1", drop=False)
    .sort_index()
)
##### Loading phenodata file #####

##### Getting all prefix from fastq.gz sample filename #####
#SAMPLES = phenodata[["Filename_R1"]].replace("_S.*", "", regex=True).to_numpy().flatten()
SAMPLES = phenodata[["Filename_R1"]].replace("_R1_001.fastq.gz", "", regex=True).to_numpy().flatten()
##### Getting all prefix from fastq.gz sample filename #####

# TODO test recover sample names via wildcards
##### Getting all prefix from folder and using glob_wildcards
#SAMPLES = glob_wildcards("../resources/reads/{sample}.fastq.gz")
##### Getting all prefix from folder and using glob_wildcards

rule all:
    input:
        ### QC results ###
        "results/qc/multiqc_report.html",
        #"results/qc/barplot_pca.png",
        #"results/qc/scatterplot_pca.png",
        ### QC results ###

        "results/counts/raw_count_matrix_ensemblid.tsv",
        expand("results/counts/intersect/{sample}.intersected.bed", sample=SAMPLES)
        #"results/assembly/trinity/Trinity.fasta"

##### Loading rules #####
include: "rules/qc.smk" # PCA plot and check others QC processes as well
include: "rules/ref.smk"
include: "rules/map-bowtie2.smk"
include: "rules/featurecounts.smk"
include: "rules/assembly_trinity.smk"
include: "rules/trim.smk"
include: "rules/nonannotated_reads.smk"
##### Loading rules #####


##### Execution #####
# snakemake --use-conda --cores 6 --latency-wait 20
#
# snakemake --dag | dot -Tsvg > dag.svg
# snakemake --dag | dot -Tpng > dag.png
##### Execution #####

##### If you face in the error message #####
# ModuleNotFoundError: No module named 'snakemake_wrapper_utils'
# You should install this module by this command line: pip install snakemake-wrapper-utils
##### If you face in the error message #####
